### ISSUE TITLE
php bin/magento catalog:images:resize does not process hidden images.

### ISSUE THREAD

**Issue by gwharton** [magento/magento2#28239]
### Preconditions (*)
1. Magento 2.4-develop

### Steps to reproduce (*)
1. Create a Product
2. Assign Image 1 to product and set to be visible on product page
3. Assign Image 2 to product and set to be hidden on product page
4. Save the product

### Expected result (*)
1. View the product page, only Image 1 should be visible
2. pub/media/catalog/cache folder should contain resized Image 1's
3. pub/media/catalog/cache folder should contain resized Image 2's
4. Clear image cache
5. Run php bin/magento catalog:images:resize
6. pub/media/catalog/cache folder should contain resized Image 1's
7. pub/media/catalog/cache folder should contain resized Image 2's

### Actual result (*)
1. Only Image 1 is visible on product page ✅
2. Image 1 is present in cache in resized states ✅
3. Image 2 is present in cache in resized states ✅
4. Cache is cleared ✅
5. Image resize command processes 1 image, not 2 ❌
6. pub/media/catalog/cache folder should contain resized Image 1's ✅
7. pub/media/catalog/cache folder does not contain resized Image 2's ❌

Images that are hidden from the product page are not processed by the image resize command. They should be. Just because they are hidden from the product page, doesn't mean they won't be used elsewhere.

Magento is written so that hiding the product image, actually sets it to disabled.

IMHO setting a product image to disabled, and setting it to be hidden on product page are two very different settings, but magento currently treats them as the same.

The confusion stems from the hidden setting being stored in the database as the field disabled. A hidden image is NOT a disabled image, and should still be processed, just not displayed on the product page, otherwise, what is the point of the hidden setting?

### Additional Information
https://github.com/magento/magento2/issues/28239#issuecomment-633611868

**Comment by hemendraajmera** [comment]
I have also mentioned this issue as https://github.com/magento/magento2/issues/28182

**Comment by gwharton** [comment]
Note, this behaviour was introduced in commit 048a2ee6b6f7215ea3168773730bff2f0ec2f01f

**Comment by gwharton** [comment]
Looks like the change was made on purpose as an optimisation.

**Comment by engcom-Delta** [comment]
HI @gwharton could you provide case where we can reuse same hidden image? Image uploaded for product cannot be used by media gallery and same image uploaded for another product has new name in database (name + increment index) 
![image](https://user-images.githubusercontent.com/51681379/82822592-b5cf9600-9eae-11ea-907e-49ab3dba876c.png)

**Comment by gwharton** [comment]
@engcom-Delta I have a separate custom written module that uploads product data to Google Shopping and Google My Business. I need to use different product images for Google feeds than on my front end, so when I create a product I upload two images, one for front end, and one hidden one for Google. My module then picks up this hidden image url and uses that on the google feeds.

The setting in Magento is named "Hide from Product Page", so seems like the perfect setting for me. And infact, it is the way that Magento works when you initially upload an image for a product. Setting it to hidden on the catalog page, DOES generate the cached images for it, regardless of its hidden state. The cached images only dissapear when you clear image cache and then regenerate using catalog:images:resize. There is an inconsistency there.

If this isn't going to be fixed for catalog:images:resize, then Magento should be changed so that it does not generate cached images for new images when they are uploaded to make it consistent, i.e cached images are not generated for hidden images, and the setting should also be changed on the frontend to "Disabled" instead of "Hidden from product page" to reflect the actual setting that is applied.

**Comment by gwharton** [comment]
In my module I call 

```
$this->imageBuilder->setProduct($product)
                    ->setImageId('product_page_main_image_gmb')
                    ->setAttributes([])
                    ->create()
                    ->getImageUrl();
```

where `product_page_main_image_gmb` is a custom image type I have defined in my view.xml.

The url returned by this call returns a 404, when the image is "Hidden from product page" as the cached images were not processed by catalog:images:resize after a images cache clear.

**Comment by gwharton** [comment]
Perhaps there should be 2 settings, "hidden" and "disabled"

**Comment by piyushbrainvire** [comment]
Due to this cart and wishlist page shows placeholder images if small/thumbnail image is set to Hide from Product page.
Is there any update solution for this?

**Comment by gwharton** [comment]
You can manually apply a patch as a workaround.

```composer require vaimo/composer-patches```

patches/composer/imageresizeall.diff
```
--- Service/ImageResize.orig.php	2020-07-17 14:15:47.350000000 +0100
+++ Service/ImageResize.php	2020-07-17 14:25:30.870000000 +0100
@@ -180,12 +180,12 @@
      */
     public function resizeFromThemes(array $themes = null): Generator
     {
-        $count = $this->productImage->getCountUsedProductImages();
+        $count = $this->productImage->getCountAllProductImages();
         if (!$count) {
             throw new NotFoundException(__('Cannot resize images - product images not found'));
         }
 
-        $productImages = $this->productImage->getUsedProductImages();
+        $productImages = $this->productImage->getAllProductImages();
         $viewImages = $this->getViewImages($themes ?? $this->getThemesInUse());
 
         foreach ($productImages as $image) {
```

composer.json
```
    "extra": {
        "magento-force": "override",
        "composer-exit-on-patch-failure": true,
        "patches": {
            "magento/module-media-storage": {
                "https://github.com/magento/magento2/issues/28239": "patches/composer/imageresizeall.diff"
            }
        }
    }

```

Works on 2.3.5-p1

**Comment by gwharton** [comment]
Perhaps a compromise would be a command line argument --process-hidden-images with the default to not process hidden images??

**Comment by Flipmediaco** [comment]
The wording "Hide from Product Page" implies the setting will only hide the image from the product page, but allowing the image to be used on the category listing page, wish list, compare, basket, any other use that is not product page... The original behaviour of only suppressing display of the image in the product page gallery is in line with the wording.

The optimisation of not generating images when "Hide from Product Page" is checked differs from the original intended purpose, which is clearly to stop supporting product images from displaying in the product page image gallery.

Many thanks to @gwharton for your patch, this works perfectly and allows the setting "Hide from Product Page" to be used as originally conceived.

**Comment by gwharton** [comment]
The catalog:images:resize routine seems to have been butchered a bit lately, but the way I see it, there are two recent changes that would have been better suited to command line arguments to make them optional, than making it work the same for everyone.

--skip-hidden-images = Do not process images marked as hidden from product page
--skip-if-exists = Do not process an image, if it already exists. (Since https://github.com/magento/magento2/pull/26801 If you change the image, but keep the filename the same, there is no way of regenerating that image, apart from clearing entire image cache)

**Comment by xi-ao** [comment]
This issue is not solved and several proposals here should be considered for the main branch.

**Comment by silkspin** [comment]
This is causing big problems for me in 2.4.1 too. I also agree with the proposals above. I have come up with my own workaround until this problem is fixed, but it might not be suitable for everyone. It involves making changes to the database. I have added "Small" to the end of my image labels ("Alt Text" in the admin) to differentiate the large product images from the smaller category page images. These Small category images have "Hide from Product Page" checked in the admin. Using MySQL I search on the labels for the ones that end with Small in the table `catalog_product_entity_media_gallery_value` and change them to disabled=0. Then I run `php bin/magento catalog:images:resize` in the CLI. After this I revert back to disabled=1 with a MySQL query and clean the cache. This prevents the resize command ignoring the hidden images when generating the image cache, and prevents showing the Small images on the product page.

phpMyAdmin
````UPDATE `catalog_product_entity_media_gallery_value` SET `disabled` = '0' WHERE `label` LIKE '%Small';````

CLI
`php bin/magento catalog:images:resize`

phpMyAdmin 
````UPDATE `catalog_product_entity_media_gallery_value` SET `disabled` = '1' WHERE `label` LIKE '%Small';````

CLI
`php bin/magento cache:clean`

All my products are configurable. Each simple product has just a Small and Thumbnail image for use in the cart / checkout. I used to have these hidden, "Hide from Product Page" checked, but now I've set them permanently to disabled=0. That means they are included when the resize command is run. I don't think this has a detrimental effect anywhere else. All of this hassle could be eradicated by just fixing the hidden / resize problem so it works how it used to.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Delta 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40921` were created

**Issue Available**: @engcom-Delta, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by BogdanGorbach** [comment]
@magento I am working on this

**Comment by BogdanGorbach** [comment]
@magento add to contributors team

**Comment by BogdanGorbach** [comment]
@magento I am working on this

**Comment by gabrieldagama** [comment]
Hi @gwharton. Thank you for your report. 
The issue has been fixed in magento/magento2#33452 by @BogdanGorbach in 2.4-develop branch
Related commit(s):
 - [c35cd1fe9fe7a909bbc6773a9ffd17d1f4480424](https://github.com/magento/magento2/commit/c35cd1fe9fe7a909bbc6773a9ffd17d1f4480424)
 - [d16ee89ee398bbc48a3e6b423fcb919658bfd369](https://github.com/magento/magento2/commit/d16ee89ee398bbc48a3e6b423fcb919658bfd369)
 - [788b2fc1ccf31ea5dd846319ac75ced214eccdd5](https://github.com/magento/magento2/commit/788b2fc1ccf31ea5dd846319ac75ced214eccdd5)

The fix will be available with the upcoming 2.4.3 release.
